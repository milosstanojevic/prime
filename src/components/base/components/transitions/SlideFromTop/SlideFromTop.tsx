import React, { FC, ReactElement, cloneElement } from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from '../types';

interface ISlideFromTop extends Omit<TransitionProps, 'timeout'> {
    children: ReactElement;
    timeout?: number;
}

const defaultStyle = {
    transitionProperty: 'opacity, transform',
    transitionTimingFunction: 'linear ease-out',
    transform: 'translateY(-100%)',
    opacity: 0
};

const transitionStyles = {
    entering: { opacity: 1, transform: 'translateY(0)', transitionDuration: '.3s' },
    entered: { opacity: 1, transform: 'translateY(0)', transitionDuration: '.3s' },
    exiting: { opacity: 0, transform: 'translateY(-100%)', transitionDuration: '.2s' },
    exited: { opacity: 0, transform: 'translateY(-100%)', transitionDuration: '.2s' }
};

export const SlideFromTop: FC<ISlideFromTop> = ({
    in: inProp = false,
    onEnter,
    onExited,
    children,
    style,
    timeout = 400
}) => (
    <Transition appear in={inProp} timeout={timeout} onEnter={onEnter} onExited={onExited}>
        {(state): ReactElement => {
            return cloneElement(children, {
                style: {
                    visibility: state === 'exited' && !inProp ? 'hidden' : undefined,
                    ...defaultStyle,
                    // @ts-ignore
                    ...transitionStyles[state],
                    ...style,
                    ...children.props.style
                }
            });
        }}
    </Transition>
);

SlideFromTop.displayName = 'SlideFromTop';
