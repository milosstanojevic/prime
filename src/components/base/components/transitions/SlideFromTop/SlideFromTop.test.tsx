import React from "react"
import { render } from "enzyme"
import { SlideFromTop } from "./SlideFromTop"

describe("<SlideFromTop />", () => {
  it("should render", () => {
    expect(render(
      <SlideFromTop>
        <span>Test</span>
      </SlideFromTop>
    )).toMatchSnapshot();
  });

  it("should render as entered", () => {
    expect(render(
      <SlideFromTop in>
        <span>Test</span>
      </SlideFromTop>
    )).toMatchSnapshot();
  });
});
