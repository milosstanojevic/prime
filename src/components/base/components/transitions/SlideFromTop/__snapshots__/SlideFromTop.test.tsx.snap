// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlideFromTop /> should render 1`] = `
<span
  style="visibility:hidden;transition-property:opacity, transform;transition-timing-function:linear ease-out;transform:translateY(-100%);opacity:0;transition-duration:.2s"
>
  Test
</span>
`;

exports[`<SlideFromTop /> should render as entered 1`] = `
<span
  style="transition-property:opacity, transform;transition-timing-function:linear ease-out;transform:translateY(-100%);opacity:0;transition-duration:.2s"
>
  Test
</span>
`;
